[12.0] Fix  issue with datamodel, recursion limit

lmignon opened the pull request:
supersedes #118 includes #117

sebastienbeau committed:
[BUG] issue with datamodel, recursion limit

lmignon committed:
[FIX] datamodel: Avoid maximum recursion

Into the datamodel constructor, we 'patch' the datamodel constructor to ensure the environment is propagated when a instance is created. Before this change, each time the factory was called, the init method was patched again. After some times, the maximum recusion limit was exceed. With this change we ensure that the patch is applied only once.

lmignon commented:
@etobella @sebastienbeau The issue is fixed without removing functionalities.  Thank you for all your investigations!

sbidoul commented on datamodel/core.py:
```diff
@@ -388,6 +391,7 @@ def __get_schema_class__(cls, **kwargs):
             return cls
 
         model.__get_schema_class__ = __get_schema_class__
+        setattr(model, "__datamodel_init_patched", object())  # noqa: B010
```
```suggestion
        setattr(model, "__datamodel_init_patched", True)  # noqa: B010
```

lmignon replied:
Thank you!

sebastienbeau commented:
@lmignon Thanks I will test it real soon !

lmignon commented:
@sebastienbeau I discovered an issue with the current implementation (and the previous one). The environment on the datamodel is not consistent.... IOW, You're not ensured that the property env on datamodel instance always returns the env from where the instance has been initialized..... I've to thing further how to fix this issue if it's possible.

etobella commented:
Not working properly, the environment is not correctly passed  to the datamodel.

Check the error using the following commit:
https://github.com/tegin/rest-framework/commit/4822c50efb575a2322059bf2cf370c234908ff83

---

I did not receive your comment on time, sorry

lmignon commented:
@etobella You're right, I just added the test on my side. The problem was already there into the previous implementation. I'll try to find a solution in the coming days...

rvalyi commented:
Hello I can confirm this fixed a real production issue I had on v12. Tested before and after, it really fixed it, thanks.

Commit:
[FIX] fix unit test w/ odoo test suite

This commit
odoo/odoo@ed831ab
introduced a backward imcompatible change
that makes simple unit tests ran through Odoo suite fail
because `doClassCleanups` is not defined.
They defined it on `TreeCase` so... there we go.

Add comment about previous fix

add comment about previous fix

lmignon committed:
[IMP] datamodels: Add readme

lmignon commented:
/ocabot merge patch
